### PR TITLE
[sdk/go] Add Contains method to PropertyPath

### PR DESCRIPTION
**Pull request by lblackstone**
# Description

Related to https://github.com/pulumi/pulumi/issues/6753

## Checklist

- [x] I have added tests that prove my fix is effective or that my feature works

- [x] I have updated the [CHANGELOG-PENDING](https://github.com/pulumi/pulumi/blob/master/CHANGELOG_PENDING.md) file with my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

File: sdk/go/common/resource/properties_path.go [hunk 1]
```diff
@@ -257,3 +257,37 @@ func (p PropertyPath) Delete(dest PropertyValue) bool {
 	return true
 
 }
+
+// Contains returns true if the receiver property path contains the other pproperty path.
+// For example, the path `foo["bar"][1]` contains the path `foo.bar[1].baz`.  The key `"*"`
+// is a wildcard which matches any string or int index at that same nesting level.  So for example,
+// the path `foo.*.baz` contains `foo.bar.baz.bam`, and the path `*` contains any path.
```
Comment (pgavlin) [inline sdk/go/common/resource/properties_path.go:264]: This comment doesn't seem accurate w.r.t the code. From what I see, `*` only works with dotted property accesses, not with brackets.

File: sdk/go/common/resource/properties_path.go [hunk 2]
```diff
@@ -257,3 +257,37 @@ func (p PropertyPath) Delete(dest PropertyValue) bool {
...
+func (p PropertyPath) Contains(other PropertyPath) bool {
+	if len(other) < len(p) {
+		return false
+	}
+
+	for i := range p {
+		pp := p[i]
+		otherp := other[i]
+
+		switch pp.(type) {
```
Comment (pgavlin) [inline sdk/go/common/resource/properties_path.go:274]:
```suggestion
		switch pp := pp.(type) {
```

File: sdk/go/common/resource/properties_path.go (hunk 1 above)
Comment (lukehoban) [reply sdk/go/common/resource/properties_path.go:264]: Good catch - I suggest we update the code to match the comment, and add a test covering this.

Comment (lblackstone) [reply sdk/go/common/resource/properties_path.go]: I added some more test cases for wildcards, and confirmed that the existing logic works as expected.

Comment (pgavlin) [reply sdk/go/common/resource/properties_path.go]: @lblackstone are you also going to update the code to match the comment, or will we take care of that in a future change?

Comment (lukehoban) [reply sdk/go/common/resource/properties_path.go]: Oh - indeed as @lblackstone noted above - the code does actually match the comment. The new tests make this clear.

**Review by lukehoban — APPROVED** [review]

**Comment by lblackstone** [comment]
It looks like the test failure is coming from https://github.com/pulumi/pulumi/pull/7049

cc @t0yv0

File: sdk/go/common/resource/properties_path.go (hunk 1 above)
Comment (pgavlin) [reply sdk/go/common/resource/properties_path.go]: Ah, I see it now! Tricky. Thanks for pointing that out.

**Review by pgavlin — APPROVED** [review]

**Comment by t0yv0** [comment]
Logged flaky https://github.com/pulumi/pulumi/issues/7361 - the failure should not block merging (esp since only fails on mac runner)

**Pull request merged**